Formally define production syntax.

Thom1729 opened the pull request:
Provide a formal definition of the production syntax. Based in part on #158.

Special productions use the `<name>` syntax to distinguish them from normal productions.

Only a few comments remain — the 1024 character limit and the c-forbidden stuff.

Because alternation is defined to be greedy, and the quantifiers are all defined using alternation, the old text describing this is no longer necessary.

Thom1729 committed:
Formally define production syntax

Thom1729 commented:
https://spec.yaml.io/pr-183

ingydotnet committed:
Markdown tool tweaks

ingydotnet commented on spec/1.2.2/spec.md:
```diff
@@ -2375,10 +2406,10 @@ convey [content] information.
 
 s-indent(0) ::=
-  /* Empty */
+  <empty>
```
Can we replace `<empty>` with `""` like in `e-scalar` ?

Thom1729 replied:
We could; it's a matter of preference. I like `<empty>` because it's explicit and because my eyes get lost in `""`.

ingydotnet replied:
How about
```
s-indent(0) ::= ""    # Empty
```

I like `""` because it's concrete, and we have one less "special"

eemeli replied:
I would prefer `<empty>`. If/as it looks like `""` would need an explanatory comment, it ought to instead have a name that avoids the issue.

ingydotnet replied:
ok

ingydotnet commented on spec/1.2.2/spec.md:
```diff
@@ -1268,6 +1259,46 @@ The right-hand view uses JSON when possible.
+  * `[lookahead ≠ term]`, which matches the empty string if `term` would not
+    match.
+  * `[lookbehind = term]`, which matches the empty string if `term` would match
+  beginning at any prior point on the line and ending at the current position.
```
Wrong indent

ingydotnet replied:
fixed

ingydotnet commented on spec/1.2.2/spec.md:
```diff
@@ -1268,6 +1259,46 @@ The right-hand view uses JSON when possible.
+  * `[lookbehind = term]`, which matches the empty string if `term` would match
+  beginning at any prior point on the line and ending at the current position.
+* A special production:
+  * `<empty>`, which matches the empty string.
```
Maybe eliminate in favor of `""`. See below.

ingydotnet commented on spec/1.2.2/spec.md:
```diff
@@ -1268,6 +1259,46 @@ The right-hand view uses JSON when possible.
+* A special production:
+  * `<empty>`, which matches the empty string.
+  * `<impossible>`, which cannot match.
+  * `<start of line>`, which matches the empty string at the beginning of a
```
Any way to define this with a lookbehind?

Thom1729 replied:
Probably, but it might be a bit complicated because we don't want to match the middle of a CRLF. Also, we'd like to eliminate the existing lookbehind (in 1.3 if not in 1.2.2) but I'm not sure we could eliminate the start-of-line symbol.

ingydotnet commented on spec/1.2.2/spec.md:
```diff
@@ -1268,6 +1259,46 @@ The right-hand view uses JSON when possible.
+    line.
+  * `<end of input>`, matches the empty string at the end of the input.
+* A parenthesized term, which matches its contents.
+* A sequence or alternation:
```
concatenation or alternation?

sequence is overloaded in spec, maybe.

ingydotnet replied:
updated

ingydotnet commented on spec/1.2.2/spec.md:
```diff
@@ -1268,6 +1259,46 @@ The right-hand view uses JSON when possible.
+
+The order of operations is parenthesization, then quantification, then
+sequence, then alternation. For example, `a b | c d` is equivalent to
+`(a b) | (c d)`.
```
I'd rather not bother with this. We can be explicit with parens in the few places we need to be.

Thom1729 replied:
I think this would add a lot of otherwise-unneeded parens. But I'll take a look at the productions.

eemeli replied:
I would have a strong preference for explicitly including the order of operations.

ingydotnet commented on spec/1.2.2/spec.md:
```diff
@@ -1268,6 +1259,46 @@ The right-hand view uses JSON when possible.
+* A quantified term:
+  * `term?`, which matches `(term | <empty>)`.
+  * `term*`, which matches `(term term* | <empty>)`.
+  * `term+`, which matches `(term term*)`.
```
What happened to greedy?

Thom1729 replied:
Alternation is defined to prefer the first branch, so these are greedy by the way they're defined. E.g. `(<empty> | term)` would be lazy.

ingydotnet replied:
I think it's worth adding a note here.

ingydotnet commented on spec/1.2.2/spec.md:
```diff
@@ -1268,6 +1259,46 @@ The right-hand view uses JSON when possible.
+sequence, then alternation. For example, `a b | c d` is equivalent to
+`(a b) | (c d)`.
+
+Whitespace in productions is not significant except inside quoted strings.
```
Do we need to say this?

Thom1729 replied:
Arguably not, but I think it's inoffensive at worst.

ingydotnet commented on spec/1.2.2/spec.md:
```diff
@@ -1268,6 +1259,46 @@ The right-hand view uses JSON when possible.
+  * `term?`, which matches `(term | <empty>)`.
+  * `term*`, which matches `(term term* | <empty>)`.
+  * `term+`, which matches `(term term*)`.
+
```
Where is `{n}` quantifier?

Thom1729 replied:
It doesn't exist anymore since s-indent was redefined.

ingydotnet replied:
:+1:

ingydotnet commented on spec/1.2.2/spec.md:
```diff
@@ -1268,6 +1259,46 @@ The right-hand view uses JSON when possible.
+term is either:
+
+* An atomic term:
+  * A quoted string (`"abc"`), which matches that sequence of characters.
```
Say quotes may use single or double form.

eemeli replied:
Is there a reason to allow for single quotes in term definitions?

ingydotnet replied:
They read cleaner and are useful for defining `'"'` cleanly.

They also have strong 'character' type semantics in some languages.

I added the concept of double quotes to not need: `'Y' 'A' 'M' 'L'` stuff but I'm not keen on collapsing to double quotes for everything just because we can.

ingydotnet commented on spec/1.2.2/spec.md:
```diff
@@ -2393,13 +2424,13 @@ The productions use the notation "**`s-indent-less-than(n)`**" and
 [#]
 s-indent-less-than(-1) ::=
-  /* Impossible */
+  <impossible>
```
Do we need impossible productions in the spec?

Thom1729 replied:
Yes, at least as currently written, for two reasons:

- n can be -1, which makes all the indentation productions impossible.
- `s-indent-less-than(0)` is impossible. We can actually eliminate `s-indent-less-than(n)` and replace it everywhere with `s-indent-less-or-equal(n+1)`, but that would mean removing a production.

In 1.3, when we fix the -1 thing, we won't need it anymore.

ingydotnet replied:
I still don't get it. I understand that the productions you added with `<impossible>` _are_ impossible.
But that's obvious from the series terminating productions next to them.

Can you show me an example of a YAML stream where a parser would reach those productions?

Thom1729 replied:
It's a consequence of the definition. If we say that *n* can be either a natural number or -1, then `s-indent(n)` should be defined for -1.

If we want to omit the impossible productions, we'll have to add a special note that for those productions, *n* may not be -1 (or zero in the case of `s-indent-less-than`).

eemeli reviewed (approved):
Looks good. Would prefer avoiding spaces within terms, hence the suggestions renaming `<start-of-line>` and `<end-of-input>`.

Thom1729 committed:
Apply suggestions from code review

Co-authored-by: Eemeli Aro <eemeli@gmail.com>

ingydotnet committed:
Various changes to production related things

I removed the `<impossible>` productions but can add them back if really
needed. Let's discuss.

ingydotnet committed:
Fix tests

ingydotnet commented on spec/1.2.2/spec.md:
```diff
@@ -1268,6 +1259,46 @@ The right-hand view uses JSON when possible.
+* A sequence or alternation:
+  * `term-one term-two`, which matches `term-one` followed by `term-two`.
+  * `term-one | term-two`, which matches the `term-one` if possible, or
+    `term-two` otherwise.
```
Agree and already refactored.

ingydotnet commented on spec/1.2.2/spec.md:
```diff
@@ -1268,6 +1259,46 @@ The right-hand view uses JSON when possible.
+  * `<impossible>`, which cannot match.
+  * `<start of line>`, which matches the empty string at the beginning of a
+    line.
+  * `<end of input>`, matches the empty string at the end of the input.
```
Thom and I both did.

ingydotnet committed:
Add a clarifying note about quantification greediness

Thom1729 commented on spec/1.2.2/spec.md:
```diff
@@ -1268,6 +1259,46 @@ The right-hand view uses JSON when possible.
+  * `term*`, which matches `(term term* | <empty>)`.
+  * `term+`, which matches `(term term*)`.
+
+> Note: Quantified terms are always greedy and do not have the backtracking
```
The quantifiers as defined are greedy, but not possessive. E.g., `"a"* "a"` would match the string `a`.

Thom1729 committed:
Limit indentation rules to n≥0

Thom1729 committed:
Reorder terms in less-than/equal productions

ingydotnet reviewed (approved):
Love it
